src/compiler/checker.ts

```diff
@@ -17663,18 +17698,20 @@ namespace ts {
                 if (usedInFunction) {
                     // mark iteration statement as containing block-scoped binding captured in some function
                     let capturesBlockScopeBindingInLoopBody = true;
-                    if (isForStatement(container) &&
-                        getAncestor(symbol.valueDeclaration, SyntaxKind.VariableDeclarationList)!.parent === container) {
-                        const part = getPartOfForStatementContainingNode(node.parent, container);
-                        if (part) {
-                            const links = getNodeLinks(part);
-                            links.flags |= NodeCheckFlags.ContainsCapturedBlockScopeBinding;
-
-                            const capturedBindings = links.capturedBlockScopeBindings || (links.capturedBlockScopeBindings = []);
-                            pushIfUnique(capturedBindings, symbol);
-
-                            if (part === container.initializer) {
-                                capturesBlockScopeBindingInLoopBody = false; // Initializer is outside of loop body
+                    if (isForStatement(container)) {
+                        const varDeclList = getAncestor(symbol.valueDeclaration, SyntaxKind.VariableDeclarationList);

```

Comment:
I made a similar change below (on line 17726) to prevent a crash that happened with a block-scoped binding that was generated by the transform of nested class expressions. When classes are in a loop and we have non-literal computed property names or private names, we need to generate a block-scoped variable to hold the computed property name or WeakMap for the private named field. Without the undefined check below, it was crashing because the block-scoped binding's valueDeclaration wasn't actually a `VariableDeclaration` but it corresponded to the property name within the class.

I noticed this code was making the same assumption (that `valueDeclaration` was a `VariableDeclaration`), but I wasn't sure how to reproduce this case. I changed it to be defensive against the `valueDeclaration` not being a `VariableDeclaration` because there are cases where that is true.